tools/src/main/scala/scala/scalanative/build/LLVM.scala

```diff
@@ -298,10 +298,22 @@ private[scalanative] object LLVM {
       buildTargetLinkOpts ++ flto ++ debugFlags ++ platformFlags ++ linkNameFlags ++ output ++ sanitizer ++ target
     }
     val paths = objectsPaths.map(_.abs)
+    val (lldStartOptions, lldEndOptions) =
+      if (config.linkingOptions.contains("-fuse-ld=lld")) {
+        // lld requires that object files are listed in the order
+        // they require each other. We don't do that so we wrap
+        // the files in --start-lib and --end-lib which consider
+        // them like they were in a .a library and links all symbols
+        // regardless of ordering
+        (List("-Wl,--start-lib"), List("-Wl,--end-lib"))
+      } else {
+        (Nil, Nil)
+      }
     // it's a fix for passing too many file paths to the clang compiler,
     // If too many packages are compiled and the platform is windows, windows
     // terminal doesn't support too many characters, which will cause an error.
-    val llvmLinkInfo = flags ++ paths ++ linkopts
+    val llvmLinkInfo =
+      flags ++ lldStartOptions ++ paths ++ lldEndOptions ++ linkopts
```

Comment:
I agree that alternative approach in #4161 might be too eager, but probably we can adopt one change added there --start-lib to make it a bit simpler and reduce ammount of allocations 

https://github.com/scala-native/scala-native/blob/092334a9d87f488ba4e93a89a759eb94155a447b/tools/src/main/scala/scala/scalanative/build/LLVM.scala#L308-L325

Reply:
I've used your suggestion and extended to `links`.
Ready for another round of review @WojciechMazur